README
simpleS - a simple server for node.js


CHANGELOG.md
## 0.9.0-alpha (in progress)
- Full rewrite using ES6 syntax and new language feature
- Support Node 8+
- Use MIT licence
- General performance improvements and bug fixes
- Removed `recache` dependency
- Moved browser client to `simples-ws` package
- Changed `connection.session` API to use Map methods
- Changed connection parsing API to use `.form()` and `.body()` methods
- Added improved hierarchic router
- Removed `host.config()`
- Removed possibility to unuse middlewares
- Removed `host.middleware()`
- Removed `host.leave()`
- Renamed `.del()` route to `.delete()` route
- Removed possibility to add multiple routes with the same listener
- Removed `host.leave()`
- Renamed ws modes to boolean `advanced` mode
- Removed `.ajax()` and `.ee()` browser API

## 0.8.8
- Fix big WebSocket frames parsing
- Fix possible server crash by sending invalid WebSocket frames

## 0.8.7
- Fix client WebSocket connection creation
- Fix dynamic HTTP host removal

## 0.8.6
- Added virtual hosts defined with wild cards, websocket host included
- Added `host.data`
- Added `connection.use()` and `connection.unuse()` methods
- Improve websocket message broadcasting
- Improved internal code reutilization

## 0.8.5
- Fixed `simples.server` not being defined
- Now port argument will overwrite port property in server or mirror options
- Removed limitation for error routes to use only 404, 405 and 500 errors
- Added possibility to trigger defined error routes using the `estatus` event of the host

## 0.8.4
- Added `start` and `stop` events for server and mirror instances
- Fixed default session store to be compliant to the specs described in the docs
- Fixed default session clean up
- Throw error when mirror is on the same port with the main server instead of silent fail
- Improve restart functional for server and mirror instances
- Added some improved tests

## 0.8.3
- Add `serve` event when the host cache is ready
- Fix `connection.cache()` max age option checking
- Fix WS channel `.broadcast()` with simple modes

## 0.8.2
- Fix incorrect HTTP host configuration

## 0.8.1
- Reveal `.connections` for WS host and channel in the public docs
- Fix streams to work properly in node >0.12 versions
- Fix rendering in WebSocket connection
- Fix middleware removing

## 0.8.0
- Added Mirrors
- No more HTTP + HTTPS functional, use Mirrors instead
- Improved Client API request piping
- Improved QS parser to force array values
- Fixed render listeners creation
- Fixed websocket connection keep alive process in node >0.12 versions
- Added `timeout` option for HTTP and WS hosts

## 0.7.6
- Add `connection.cache()` method
- Fix importer to be optional in host routing when the listener is a string

## 0.7.5
- Update recache dependency for bug fixing

## 0.7.4
- Expose connection session to middlewares
- Added `simples.server()` for consistency
- `connection.keep()` will always set 0 timeout for invalid timeout value
- Added options to `host.serve()` for configuring the cache
- Moved cache and mime to external dependencies

## 0.7.3
- Improved template engine capabilities
- Fixed WebSocket limit for received data
- Fixed possible crash on WebSocket errors
- Small docs fixes

## 0.7.2
- Fixed implementation of routes with functions as importer parameter
- Improved performance for session keys generation

## 0.7.1
- Improved ws pong frame creation
- Improved performance for broadcasting data in websocket host and channel with filter
- Changed the way errors are emitted in non-TTY environment
- Improved routing process by removing an unnecessary checking step
- Moved routing process to be after all middlewares are executed
- Fixed some bugs related to CORS and websocket channel closing

## 0.7.0
- Add support for node.js 0.12 and io.js 1.x
- Implemented `simples.client`
- Disabled response compression by default
- Modified `connection.ip` to be `net.socket.address()` content
- Added `connection.data`
- Removed `referers` from HTTP host configuration as it's not so useful
- Added more options for CORS in host configuration
- Use plain parser when no other parser is valid for request parsing
- Simplified parsing API for `json` and `urlencoded` data
- Improved internal request routing and static file caching
- Replaced `type` option, `raw` and `advanced` modes from WebSocket configuration with `binary`, `text` and `object` modes
- Added separated accepted origins for WebSocket configuration
- Improved client-side code
- Improved examples
- Fixed documentation errors
- Bug fixes

## 0.6.6
- Fixed and improved documentation to be up-to-date with the code
- Rewritten the client-side to wrap the code inside a closure
- Improved the client-side code
- Improved the search of the dynamic routes

## 0.6.5
- Fixed documentation about routes with render and importer
- Improved requests routing
- Fixed crash because of cache.destroy()
- Fixed crash because of ws socket receiving null data

## 0.6.4
- Improved server's commands execution
- Improved some internal functional related to websockets
- Improved connection preparation
- Fixed server crash because of possible incorrect request body
- Fixed incorrect cache reading

## 0.6.3
- Fixed HTTPS content serving
- Implement a better synchronization between HTTP and HTTPS servers

## 0.6.2
- Fixed routes which rendered views
- Fixed cache root reading
- Added importer functional for `connection.render()`
- Fixed channels container structure
- Improved internal structure
- Documentation fixes

## 0.6.1
- Fixed default session store crash
- Fixed `Last-Modified` header to have value compliant the HTTP standard
- Added `Last-Modified` header for cached directories
- Reworked the server instance creation
- Some documentation fixes

## 0.6.0
- Improved cookies and langs parsers
- Revealed `connection.request` and `connection.response` in the public docs
- Removed session for static content
- Implemented `connection.log()`
- Removed `host.log` in favor of `connection.log`

lib/http.js
'use strict';

var cache = require('../utils/cache'),
	fs = require('fs'),
	http = require('http'),
	https = require('https'),
	httpConnection = require('./http/connection'),
	qs = require('querystring'),
	stream = require('stream'),
	url = require('url'),
	utils = require('../utils/utils'),
	wsUtils = require('./ws');

// Default callback for "Not Found"
function e404(connection) {
	connection.end('"' + connection.url.path + '" Not Found');
}

// Default callback for "Method Not Allowed"
function e405(connection) {
	connection.end('"' + connection.method + '" Method Not Allowed');
}

// Default callback for "Internal Server Error"
function e500(connection) {
	connection.end('"' + connection.url.path + '" Internal Server Error');
}

// Add listeners to the server
function httpAddListeners(server) {
	server.on('error', function (error) {
		this.parent.busy = false;
		this.parent.started = false;
		throw new Error('simpleS: Server error > ' + error.message);
	}).on('release', function (callback) {
		this.parent.busy = false;

		// Call the callback when the server is free
		if (callback) {
			callback.call(this.parent);
		}
	}).on('upgrade', wsUtils.wsRequestListener);
}

// Check if the referer header is accepted by the host
function refers(host, connection) {

	var accepted = true,
		referer,
		request = connection.request;

	if (request.headers.referer && host.conf.referers.length) {
		referer = url.parse(request.headers.referer).hostname;

		if (host.conf.referers.indexOf(referer) < 0) {
			accepted = host.conf.referers[0] === '*';
		} else {
			accepted = host.conf.referers[0] !== '*';
		}
	}

	return accepted || referer === connection.host;
}

// Returns the advanced route if found
function findAdvancedRoute(routes, location) {

	var param,
		params = {},
		route,
		value;

	['all', 'del', 'get', 'post', 'put'].some(function (verb) {
		return Object.keys(routes[verb].advanced).some(function (element) {
			var locationIndex = 0,
				routeIndex = 0;

			while (element.charAt(routeIndex)) {
				if (element.charAt(routeIndex) === location.charAt(locationIndex)) {
					routeIndex++;
					locationIndex++;
				} else if (element.charAt(routeIndex) === ':') {
					param = '';
					value = '';
					routeIndex++;

					while (element.charAt(routeIndex) && element.charAt(routeIndex) !== '/') {
						param += element.charAt(routeIndex);
						routeIndex++;
					}

					while (location.charAt(locationIndex) && location.charAt(locationIndex) !== '/') {
						value += location.charAt(locationIndex);
						locationIndex++;
					}

					params[param] = value;

					if (routeIndex === element.length) {
						route = routes[verb].advanced[element];
					}
				} else {
					params = {};
					route = null;
					break;
				}
			}

			return route;
		});
	});

	return {
		params: params,
		route: route
	};
}

// Parse data sent via POST method
function parsePOST(connection) {

	var boundary,
		boundaryLength,
		buffer = '',
		content = connection.request.headers['content-type'],
		contentType,
		currentChar = content.charAt(0),
		filecontent,
		filename,
		index = 0,
		name,
		POSTquery,
		tempIndex,
		type;

	while (currentChar === ' ') {
		index++;
		currentChar = content.charAt(index);
	}
	if (!currentChar) {
		return;
	}
	contentType = '';
	while (currentChar && currentChar !== ' ' && currentChar !== ';') {
		contentType += currentChar;
		index++;
		currentChar = content.charAt(index);
	}
	if (contentType === 'multipart/form-data') {
		while (currentChar === ' ' || currentChar === ';') {
			index++;
			currentChar = content.charAt(index);
		}
		if (!currentChar) {
			return;
		}
		while (currentChar && currentChar !== ' ' && currentChar !== '=') {
			buffer += currentChar;
			index++;
			currentChar = content.charAt(index);
		}
		if (buffer !== 'boundary' || !currentChar) {
			return;
		}
		while (currentChar === ' ' || currentChar === '=') {
			index++;
			currentChar = content.charAt(index);
		}
		if (!currentChar) {
			return;
		}
		boundary = '';
		while (currentChar) {
			boundary += currentChar;
			index++;
			currentChar = content.charAt(index);
		}
		if (!boundary) {
			return;
		}
		content = connection.body;
		index = 0;
		currentChar = content.charAt(index);
		boundaryLength = boundary.length;
		while (content.substr(index, 4 + boundaryLength) === '--' + boundary + '\r\n') {
			index += 4 + boundaryLength;
			index = content.indexOf('name="', index) + 6;
			if (index < 6) {
				break;
			}
			name = content.substring(index, content.indexOf('"', index));
			index += name.length + 1;
			tempIndex = content.indexOf('filename="', index) + 10;
			filename = undefined;
			if (tempIndex >= 10) {
				filename = content.substring(tempIndex, content.indexOf('"', tempIndex));
				index += 11 + filename.length;
				index = content.indexOf('Content-Type: ', index) + 14;
				type = content.substring(index, content.indexOf('\r', index));
				index += type.length;
			}
			index += 4;
			filecontent = '';
			while (content.substr(index, 4 + boundaryLength) !== '\r\n--' + boundary) {
				filecontent += currentChar;
				index++;
				currentChar = content.charAt(index);
			}
			if (filename) {
				connection.files[name] = {
					content: filecontent,
					filename: filename,
					type: type
				};
			} else {
				connection.query[name] = filecontent;
			}
			index += 2;
		}
	} else {
		POSTquery = qs.parse(connection.body);
		Object.keys(POSTquery).forEach(function (element) {
			connection.query[element] = POSTquery[element];
		});
	}
}

// Routes all the requests
function routing(host, connection) {

	var del,
		file,
		found,
		get,
		log,
		post,
		put,
		origin,
		request = connection.request,
		requestURL = connection.path.substr(1),
		response = connection.response,
		route,
		routes = host.routes;

	// Callback for a static file route
	function staticRoute(connection) {

		var index,
			mtime = file.stats.mtime.valueOf(),
			rstream;

		// Set the last modified time and the content type of the response
		connection.header('Last-Modified', mtime);
		connection.type(requestURL.substr(requestURL.lastIndexOf('.') + 1));

		// Check if modification time coincides on the client and on the server
		if (Number(connection.request.headers['if-modified-since']) === mtime) {
			connection.response.statusCode = 304;
			connection.end();
		} else {
			index = 0;
			rstream = new stream.Readable();
			rstream._read = function () {
				if (file.content.length > index + 16384) {
					this.push(file.content.slice(index, 16384));
					index += 16384;
				} else {
					this.push(file.content.slice(index));
					this.push(null);
				}
			};
			rstream.pipe(connection);
		}
	}

	del = request.method === 'DELETE';
	get = request.method === 'GET' || request.method === 'HEAD';
	post = request.method === 'POST';
	put = request.method === 'PUT';

	// Check for CORS requests
	if (request.headers.origin) {

		// Check if the origin is accepted
		if (exports.accepts(host, request)) {
			origin = request.headers.origin;
		} else {
			origin = connection.protocol + '://' + connection.host;
		}

		// Set CORS response headers
		response.setHeader('Access-Control-Allow-Credentials', 'True');
		response.setHeader('Access-Control-Allow-Headers', 'Content-Type');
		response.setHeader('Access-Control-Allow-Methods', 'DELETE,GET,HEAD,POST,PUT');
		response.setHeader('Access-Control-Allow-Origin', origin);

		// End the response
		if (request.method === 'OPTIONS') {
			response.end();
			return;
		}
	}

	// Find the route
	if ((del || get || post || put) && routes.all.simple[requestURL]) {
		route = routes.all.simple[requestURL];
	} else if (del && routes.del.simple[requestURL]) {
		route = routes.del.simple[requestURL];
	} else if (get && routes.get.simple[requestURL]) {
		route = routes.get.simple[requestURL];
	} else if (post && routes.post.simple[requestURL]) {
		route = routes.post.simple[requestURL];
	} else if (put && routes.put.simple[requestURL]) {
		route = routes.out.simple[requestURL];
	} else if (del || get || post || put) {

		// Find the advanced route
		found = findAdvancedRoute(routes, requestURL);

		// Populate the connection parameters and use the selected route
		if (found.route) {
			connection.params = found.params;
			route = found.route;
		}

		// Check for static files routes
		if (!route && get && refers(host, connection)) {

			// Get cached files
			file = host.cache.read(requestURL);

			// Return the route
			if (file && file.stats.isDirectory()) {
				route = host.routes.serve;
			} else if (file) {
				route = staticRoute;
			} else {
				response.statusCode = 404;
				route = routes.error[404];
			}
		} else if (!route) {
			response.statusCode = 404;
			route = routes.error[404];
		}
	} else {
		response.statusCode = 405;
		response.setHeader('Allow', 'DELETE,GET,HEAD,POST,PUT');
		route = routes.error[405];
	}

	// Listen for internal errors
	connection.on('error', function (error) {
		console.error('\nsimpleS: Internal Server Error > ' + this.url.href);
		console.error(error.stack + '\n');
		response.statusCode = 500;

		// Try to apply the route for error 500
		try {
			routes.error[500].call(host, this);
		} catch (stop) {
			console.error('\nsimpleS: Can not apply route for error 500');
			console.error(stop.stack + '\n');
		}
	});

	// Check if any logger is defined
	if (host.logger) {

		log = {};

		Object.keys(connection).filter(function (attribute) {
			return typeof connection[attribute] !== 'function';
		}).forEach(function (attribute) {
			log[attribute] = connection[attribute];
		});

		log = host.logger(log);

		// Check if the logger has defined a result
		if (typeof log !== 'undefined') {
			console.log(log);
		}
	}

	// Vulnerable code handling
	try {
		if (typeof route === 'string') {
			connection.render(route);
		} else {
			if (file && file.stats.isDirectory()) {
				route.call(host, connection, Object.keys(file.files).map(function (element) {
					return {
						name: element,
						stats: file.files[element].stats
					};
				}));
			} else {
				route.call(host, connection);
			}
		}
	} catch (error) {
		connection.emit(error);
	}
}

// Populate the body of the requet and route requests
function httpPopulateBody(host, connection) {

	var body = new Buffer(0),
		request = connection.request;

	// Check for too big request body
	function limitLength(length, callback) {
		if ((length | 0) > host.conf.limit) {
			console.error('\nsimpleS: Request Entity Too Large\n');
			request.destroy();
		} else if (callback) {
			callback();
		}
	}

	// Listener for request end event
	function onEnd() {

		// Convert buffer data to UTF-8 encoded string
		connection.body = body.toString();

		// Populate the files and the query of POST requests
		if (this.method === 'POST' && this.headers['content-type']) {
			parsePOST(connection);
		}

		// Route the requests
		routing(host, connection);
	}

	// Listener for request readable event
	function onRead() {

		var data = this.read() || new Buffer(0),
			length = body.length + data.length;

		// Concatenate body chunks
		body = utils.buffer(body, data, length);

		// Check for request body limit
		limitLength(body.length);
	}

	// Listen for incoming data
	limitLength(request.headers['content-length'], function () {
		request.on('readable', onRead).on('end', onEnd);
	});
}

// HTTP request listener
function httpRequestListener(request, response) {

	var connection,
		host;

	// Check if host is provided by the host header
	if (request.headers.host) {
		host = this.parent.hosts[request.headers.host.split(':')[0]];
	}

	// Get the main host if the other one does not exist or is inactive
	if (!host || !host.active) {
		host = this.parent.hosts.main;
	}

	// Create the HTTP connection
	connection = new httpConnection(host, request, response);

	// Set the keep alive timeout of the request socket to 5 seconds
	request.connection.setTimeout(5000);

	// Set the default content type to HTML and charset UTF-8
	response.setHeader('Content-Type', 'text/html;charset=utf-8');

	// Route the request and parse them if needed
	if (request.method === 'GET' || request.method === 'HEAD') {
		routing(host, connection);
	} else {
		httpPopulateBody(host, connection);
	}
}

// Read the provide files an return the files object in the callback
function readFiles(files, callback) {

	var count = 0;

	// Check if files are read
	function onEnd() {
		count--;

		// If all files are read provide the files to the callback
		if (count === 0) {
			callback(files);
		}
	}

	// Loop through the files object
	Object.keys(files).filter(function (element) {
		return ['cert', 'key', 'pfx'].indexOf(element) >= 0;
	}).forEach(function (element) {

		var data,
			length,
			location = files[element];

		// Increase the count of the files
		count++;

		// Prepare the buffer
		files[element] = new Buffer(0);

		// Create the readable stream
		fs.ReadStream(location).on('error', function () {
			throw new Error('simpleS: Can not read "' + location + '"');
		}).on('readable', function () {
			data = this.read() || new Buffer(0);
			length = files[element].length + data.length;
			files[element] = utils.buffer(files[element], data, length);
		}).on('end', onEnd);
	});
}

// Generate empty routes
exports.httpDefaultRoutes = function () {

	return {
		all: {
			advanced: [],
			simple: {}
		},
		del: {
			advanced: [],
			simple: {}
		},
		error: {
			404: e404,
			405: e405,
			500: e500
		},
		get: {
			advanced: [],
			simple: {}
		},
		post: {
			advanced: [],
			simple: {}
		},
		put: {
			advanced: [],
			simple: {}
		},
		serve: null
	};
};

// Create the HTTP or the HTTPS server
exports.createServer = function (options, callback) {

	var files,
		server;

	// Create the HTTPS server
	function httpsServer(options) {

		// Prepare the HTTP and the HTTPS servers
		var aux = http.Server(httpRequestListener);
		server = https.Server(options, httpRequestListener);

		// Catch the errors in the auxiliary HTTP server
		aux.on('error', function (error) {
			server.emit('error', error);
		}).on('upgrade', wsUtils.wsRequestListener);

		// Manage the HTTP server depending on HTTPS events
		server.on('open', function () {
			aux.listen(80);
		}).on('close', function () {
			aux.close();
		});

		// Add the listeners for the server
		httpAddListeners(server);

		// Send the server object through the callback
		callback(server);
	}

	// Check the options and read the files for the HTTPS server
	if (options && (options.cert && options.key || options.pfx)) {
		readFiles(options, httpsServer);
	} else if (options) {
		throw new Error('simpleS: Invalid data for the HTTPS server');
	} else {
		server = http.Server(httpRequestListener);
		httpAddListeners(server);
		callback(server);
	}
};